src/dm/DispatchManagerActions.cc

```diff
@@ -2249,7 +2303,750 @@ int DispatchManager::live_updateconf(std::unique_ptr<VirtualMachine> vm,
+            error_str = oss.str();
+
+            return -1;
+        }
+
+        set<int> sgs;
+        nic->get_security_groups(sgs);
+
+        if (sgs.find(sgid) == sgs.end())
+        {
+            oss << "VM " << vid << " NIC " << nicid << " doesn't contain SG " << sgid;
+            error_str = oss.str();
+
+            return -1;
+        }
+
+        // ----------------------------------------------------------------------
+        // Update VM data
+        // - NIC to remove SG
+        // - Template with remaining SG rules (could be empty)
+        // ----------------------------------------------------------------------
+        nic->remove_security_group(sgid);
+
+        vm->remove_security_group(sgid);
+
+        sgs.clear();
+        vm->get_security_groups(sgs);
+
+        if (sgs.find(sgid) != sgs.end())
+        {
+            sgpool->get_security_group_rules(-1, sgid, sg_rules);
+
+            vm->add_template_attribute(sg_rules);
+        }
+
+        vmpool->update(vm.get());
+
+        if ( is_update )
+        {
+            vmm->updatesg(vm.get(), sgid);
+        }
+    }
+    else
+    {
+        oss << "Could not attach SG to VM " << vid << ", VM does not exist";
+        error_str = oss.str();
+
+        return -1;
+    }
+
+    if (auto sg = sgpool->get(sgid))
+    {
+        sg->del_vm(vid);
+
+        if (!sg_rules.empty())
+        {
+            // The SG is in other NIC, keep the VM in the SG list
+            if ( is_update )
+            {
+                sg->add_updating(vid);
+            }
+            else
+            {
+                sg->add_vm(vid);
+            }
+        }
+
+        sgpool->update(sg.get());
+    }
+
+    return 0;
+}
+
+/* -------------------------------------------------------------------------- */
+/* -------------------------------------------------------------------------- */
+
+int DispatchManager::backup(int vid, int backup_ds_id, bool reset,
+                            const RequestAttributes& ra, string& error_str)
+{
+    ostringstream oss;
+
+    auto vm = vmpool->get(vid);
+
+    if ( vm == nullptr )
+    {
+        oss << "Could not create a new backup for VM " << vid
+            << ", VM does not exist";
+        error_str = oss.str();
+
+        return -1;
+    }
+
+    // -------------------------------------------------------------------------
+    // Set BACKUP state
+    // -------------------------------------------------------------------------
+    VirtualMachine::VmState state = vm->get_state();
+
+    if ( vm->backups().active_flatten() )
+    {
+        oss << "Could not create a new backup for VM " << vid
+            << ", consolidating backup increments";
+        error_str = oss.str();
+
+        return -1;
+    }
+
+    switch (state)
+    {
+        case VirtualMachine::ACTIVE:
+            if (vm->get_lcm_state() != VirtualMachine::RUNNING)
+            {
+                oss << "Could not create a new backup for VM " << vid
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

A cross-site scripting (XSS) vulnerability was detected. User-controlled data is being output without proper sanitization, allowing an attacker to inject malicious scripts. This could lead to session hijacking, malware installation, or phishing attacks. Always sanitize and encode user inputs before including them in the output, using context-appropriate encoding methods. Learn More - https://cwe.mitre.org/data/definitions/79.html.